src/goto-programs/validate_goto_model.cpp

```diff
@@ -0,0 +1,251 @@
+/*******************************************************************\
+
+Module: Validate Goto Programs
+
+Author: Diffblue
+
+Date: Oct 2018
+
+\*******************************************************************/
+#include "validate_goto_model.h"
+
+#include <set>
+
+#include "goto_functions.h"
+#include <util/invariant.h>
+
+class validate_goto_modelt
+{
+public:
+  using function_mapt = goto_functionst::function_mapt;
+
+  validate_goto_modelt(
+    const goto_functionst &goto_functions,
+    const validation_modet vm)
+    : vm{vm}, function_map{goto_functions.function_map}
+  {
+  }
+
+  void
+  do_goto_program_checks(goto_model_validation_optionst validation_options);
+
+private:
+  /// Check the goto_program has an entry point
+  ///
+  /// NB goto-cc -c will generate goto-programs without entry-points
+  /// until they are linked.
+  void entry_point_exists();
+
+  /// Check that no functions calls via function pointer are present
+  void function_pointer_calls_removed();
+
+  /// Check returns have been removed
+  ///
+  /// NB Calls via function pointer must have been removed already when
+  /// removing returns, thus enabling this check also should enable the check
+  /// that all calls via function pointer have been removed
+  ///
+  /// Sub-checks are:
+  /// - no return statements in any of the functions
+  /// - lhs of every \ref code_function_callt instruction is nil
+  /// - all return types are void (of both calls and functions themselves)
+  void check_returns_removed();
+
+  /// Check that for all
+  /// 1. functions that are called or
+  /// 2. functions of which the address is taken,
+  /// a corresponding entry exists in the function_map
```

Comment:
Does this render properly in Doxygen? I'd be worried that we end up with "Check that for all 1" or "Check that for all" as brief description.

Reply:
Ah no - it did not render properly.  "Check that for all" was the brief description.

This fixes it:
  /// Check that for all:
  /// -# functions that are called or
  /// -# functions of which the address is taken
  /// .
  /// a corresponding entry exists in the function_map

nb the "." is required so that it shows as so:
![cprover__anonymous_namespace validate_goto_model_cpp __validate_goto_modelt_class_reference](https://user-images.githubusercontent.com/33834914/52866313-4b219d80-3136-11e9-9b49-49c4958b82e0.jpg)

a "\<br\>" or "\n" don't cut it:
(below is with "\<br\>". \n elides the last line entirely)

![cprover__anonymous_namespace validate_goto_model_cpp __validate_goto_modelt_class_reference 2](https://user-images.githubusercontent.com/33834914/52866714-2c6fd680-3137-11e9-9b64-2b765d76c964.jpg)